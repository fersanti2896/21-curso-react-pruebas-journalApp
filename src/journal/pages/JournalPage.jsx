import { useDispatch, useSelector } from 'react-redux';
import { AddOutlined } from '@mui/icons-material';
import { IconButton } from '@mui/material';
import { JournalLayout } from '../layout/JournalLayout';
import { NoteView, NothingSelectedView } from '../views';
import { startNewNote } from '../../store/journal';

export const JournalPage = () => {
    const dispatch = useDispatch();
    const { isSaving, active } = useSelector( state => state.journal );

    const onNewNote = () => {
        dispatch( startNewNote() );
    }

    return (
        <>
            <JournalLayout>
                { !!active ? <NoteView /> : <NothingSelectedView /> }

                <IconButton disabled={ isSaving }
                            onClick={ onNewNote }
                            size='large'
                            sx={{ backgroundColor: 'error.main', 
                                  bottom: 50,
                                  color: 'white',
                                  ':hover': { backgroundColor: 'error.main', opacity: 0.9 },
                                  position: 'fixed',
                                  right: 50 }}>
                    <AddOutlined sx={{ fontSize: 30 }}/>
                </IconButton>
            </JournalLayout>
        </>
    )
}
