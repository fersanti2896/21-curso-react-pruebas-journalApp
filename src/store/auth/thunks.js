import { async } from "@firebase/util";
import { loginWithEmailPassword, logoutFirebase, registerUserWithEmail, singWithGoogle } from "../../firebase/providers";
import { loadNotes } from "../../helpers";
import { clearNotesLogout, setNotes } from "../journal";
import { checkingCredentials, login, logout } from "./"

export const checkingAuthentication = ( email, password ) => {
    return async( dispatch ) => {
        dispatch( checkingCredentials() );
    }
}

export const startGoogleSingIn = () => {
    return async( dispatch ) => {
        dispatch( checkingCredentials() );

        /* Se llama la funci贸n de Firebase para la autenticaci贸n con Google */
        const result = await singWithGoogle();
        
        if( !result.ok ) return dispatch( logout( result.errorMessage ) );

        dispatch( login( result ) )
    }
}

export const startCreateUserWithPassword = ({ email, password, displayName }) => {
    return async( dispatch ) => {
        dispatch( checkingCredentials() );

        /* Se llama la funci贸n de Firebase para la autenticaci贸n por JournalApp */
        const { ok, uid, photoURL, errorMessage } = await registerUserWithEmail({ email, password, displayName });

        if( !ok ) return dispatch(logout({ errorMessage, ok }))

        dispatch( login({ uid, displayName, email, photoURL, ok }) );
    }
}

export const startLoginEmailPassword = ({ email, password }) => {
    return async( dispatch ) => {
        dispatch( checkingCredentials() );

        const { ok, uid, displayName, photoURL, errorMessage } = await loginWithEmailPassword({ email, password });
        
        if( !ok ) return dispatch(logout({ errorMessage, ok }))

        dispatch( login({ uid, displayName, email, photoURL, ok }) );
    }
}

export const startLogout = () => {
    return async( dispatch ) => {
        await logoutFirebase();

        dispatch( clearNotesLogout() )
        dispatch( logout({ }) );
    }
}

export const startLoadingNotes = () => {
    return async( dispatch, getState ) => {
        const { uid } = getState().auth;

        if( !uid ) throw new Error('El UID del usuario no existe.');

        const notes = await loadNotes( uid );

        dispatch( setNotes(notes) );
    }
}