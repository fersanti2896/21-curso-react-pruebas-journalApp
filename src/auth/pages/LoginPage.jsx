import { useEffect, useMemo, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Link as RouterLink } from 'react-router-dom';
import { Google } from '@mui/icons-material';
import { Alert, Button, Grid, Link, TextField, Typography } from '@mui/material';
import { AuthLayout } from '../layout/AuthLayout';
import { useForm } from '../../hooks';
import { startGoogleSingIn, startLoginEmailPassword } from '../../store/auth';

const fromData = {
    email: '',
    password: ''
}

const formValidations = {
    email: [ ( value ) => value.includes('@'), 'El correo debe tener un @.'],
    password: [ ( value ) => value.length >= 6, 'La contraseña debe tener más de 6 caracteres.']
}

export const LoginPage = () => {
    const dispatch = useDispatch();
    const { status, errorMessage } = useSelector( state => state.auth );
    const [ formSubmitted, setFormSubmitted ] = useState( false );
    
    const { formState, email, password, onInputChange,
            isFormValid,  emailValid, passwordValid } = useForm( fromData, formValidations );  
    
    const isAuthenticating = useMemo(() => status === 'checking', [ status ]);

    const onSubmit = ( event ) => {
        event.preventDefault();
        setFormSubmitted( true );

        if( !isFormValid ) return;

        dispatch( startLoginEmailPassword( formState ) );
    }

    const onGoogleSingIn = () => {
        dispatch( startGoogleSingIn() );
    }   

    return (
        <>
            <AuthLayout title='Iniciar Sesión'>
                <form   aria-label='submit-form'
                        className='animate__animated animate__fadeIn animate__faster'
                        onSubmit={ onSubmit }>
                    <Grid container>
                        <Grid item sx={{ mb: 2 }} xs={ 12 } >
                            <TextField  error={ !!emailValid && formSubmitted }
                                        label='Correo'
                                        helperText={ emailValid }
                                        name="email"
                                        placeholder='correo@google.com' 
                                        onChange={ onInputChange }
                                        type='email'
                                        value={ email }
                                        fullWidth />
                        </Grid>

                        <Grid item sx={{ mb: 2 }} xs={ 12 }>
                            <TextField  error={ !!passwordValid && formSubmitted }
                                        inputProps={{
                                            'data-testid': 'password'        
                                        }}
                                        helperText={ passwordValid }
                                        name='password' 
                                        placeholder='Contraseña'
                                        onChange={ onInputChange } 
                                        type='password'
                                        value={ password }
                                        fullWidth />
                        </Grid>

                        <Grid container spacing={ 2 } sx={{ mb: 2 }}>
                            <Grid display={ !!errorMessage ? '' : 'none' } item xs={ 12 }>
                                <Alert severity='error'>{ errorMessage }</Alert>
                            </Grid>

                            <Grid item xs={ 12 } sm={ 6 }>
                                <Button disabled={ isAuthenticating }
                                        fullWidth
                                        type='submit'
                                        variant='contained'>
                                <Typography>Login</Typography>
                                </Button>
                            </Grid>

                            <Grid item xs={ 12 } sm={ 6 }>      
                                <Button aria-label='google-btn'
                                        disabled={ isAuthenticating }
                                        fullWidth 
                                        onClick={ onGoogleSingIn }
                                        variant='contained'>
                                    <Google />
                                    <Typography sx={{ ml: 1 }}>Google</Typography>
                                </Button>
                            </Grid>
                        </Grid>

                        <Grid container direction='row' justifyContent='end'>
                            <Typography sx={{ mr: 1 }}>¿No tienes cuenta?</Typography>
                            <Link color='inherit' 
                                    component={ RouterLink } 
                                    to='/auth/register'>
                                Crear una cuenta
                            </Link>
                        </Grid>
                    </Grid>
                </form>
           </AuthLayout>
            
        </>
    )
}
