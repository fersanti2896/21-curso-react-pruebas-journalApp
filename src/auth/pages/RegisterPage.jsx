import { useMemo, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Link as RouterLink } from 'react-router-dom';
import { Alert, Button, Grid, Link, TextField, Typography } from '@mui/material';

import { AuthLayout } from '../layout/AuthLayout';
import { useForm } from '../../hooks';
import { startCreateUserWithPassword } from '../../store/auth';

const formData = {
    email: '',
    password: '',
    displayName: ''
}

const formValidations = {
    email: [ ( value ) => value.includes('@'), 'El correo debe tener un @.'],
    password: [ ( value ) => value.length >= 6, 'La contraseña debe tener más de 6 caracteres.'],
    displayName: [ ( value ) => value.length >= 2, 'El nombre es obligatorio.']
}

export const RegisterPage = () => {
    const dispatch = useDispatch();
    const [ formSubmitted, setFormSubmitted ] = useState( false );
    const { status, errorMessage } = useSelector( state => state.auth );
    const isCheckingAuthentication = useMemo(() => status === 'checking', [ status ]);

    const { 
        formState, displayName, email, password, onInputChange, 
        isFormValid, displayNameValid, emailValid, passwordValid
    } = useForm( formData, formValidations );  

    const onSubmit = ( event ) => {
        event.preventDefault();
        setFormSubmitted( true );

        if( !isFormValid ) return;

        dispatch( startCreateUserWithPassword( formState ) );
    }

    return (
        <>
            <AuthLayout title='Crear Cuenta'>
                <form className='animate__animated animate__fadeIn animate__faster'
                      onSubmit={ onSubmit }>
                    <Grid container>
                        <Grid item sx={{ mb: 2 }} xs={ 12 } >
                            <TextField  label='Nombre Completo'
                                        error={ !!displayNameValid && formSubmitted }
                                        helperText={ displayNameValid }
                                        name='displayName' 
                                        placeholder='Tu nombre' 
                                        onChange={ onInputChange }
                                        type='text'
                                        fullWidth
                                        value={ displayName } />
                        </Grid>

                        <Grid item sx={{ mb: 2 }} xs={ 12 } >
                            <TextField  label='Correo'
                                        error={ !!emailValid && formSubmitted }
                                        helperText={ emailValid }
                                        name='email' 
                                        placeholder='google@prueba.com'
                                        onChange={ onInputChange } 
                                        type='text'
                                        fullWidth 
                                        value={ email } />
                        </Grid>

                        <Grid item sx={{ mb: 2 }} xs={ 12 }>
                            <TextField  label='Contraseña'
                                        error={ !!passwordValid && formSubmitted }
                                        helperText={ passwordValid }
                                        name='password'
                                        placeholder='Contraseña'
                                        onChange={ onInputChange } 
                                        type='password'
                                        fullWidth
                                        value={ password } />
                        </Grid>

                        <Grid container spacing={ 2 } sx={{ mb: 2 }}>
                            <Grid display={ !!errorMessage ? '' : 'none' } item xs={ 12 }>
                                <Alert severity='error'>{ errorMessage }</Alert>
                            </Grid>

                            <Grid item xs={ 12 }>
                                <Button disabled={ isCheckingAuthentication }
                                        fullWidth 
                                        type='submit'
                                        variant='contained'>
                                <Typography>Crear Cuenta</Typography>
                                </Button>
                            </Grid>
                        </Grid>

                        <Grid container direction='row' justifyContent='end'>
                            <Typography sx={{ mr: 1 }}>¿Ya tienes cuenta?</Typography>
                            <Link   color='inherit' 
                                    component={ RouterLink } 
                                    to='/auth/login'>
                                Ingresar
                            </Link>
                        </Grid>
                    </Grid>
                </form>
           </AuthLayout>
            
        </>
    )
}
