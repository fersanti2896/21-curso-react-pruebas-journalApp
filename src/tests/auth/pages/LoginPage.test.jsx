import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import { configureStore } from '@reduxjs/toolkit';
import { fireEvent, render, screen } from '@testing-library/react';

import { LoginPage } from '../../../auth/pages';
import { authSlice } from '../../../store/auth';
import { notAuthenticatedState } from '../../fixtures/authFixtures';

const mockStartGoogleSignIn = jest.fn();
const mockStartLoginEmailPassword = jest.fn();

jest.mock('../../../store/auth/thunks', () => ({
    startGoogleSingIn: () => mockStartGoogleSignIn,
    startLoginEmailPassword: ({ email, password }) => {
        return () => mockStartLoginEmailPassword({ email, password })
    }
}));

jest.mock('react-redux', () => ({
    ...jest.requireActual('react-redux'),
    useDispatch: () => (fn) => fn()
}))

const store = configureStore({
    reducer: {
        auth: authSlice.reducer
    },
    preloadedState: {
        auth: notAuthenticatedState
    }
});

describe('Pruebas en <LoginPage />', () => { 
    beforeEach(() => jest.clearAllMocks());

    test('Debe de mostrar el componente correctamente.', () => { 
        render( 
            <Provider  store={ store }>
                <MemoryRouter>
                    <LoginPage />
                </MemoryRouter>
            </Provider>
        );

        expect( screen.getAllByText('Login').length ).toBeGreaterThanOrEqual(1);
    });

    test('Debe hacer match con el snapshot', () => { 
        const { container } = render(
            <Provider  store={ store }>
                <MemoryRouter>
                    <LoginPage />
                </MemoryRouter>
            </Provider>
        );

        expect( container ).toMatchSnapshot();
    });

    test('El botÃ³n de Google debe de llamar a startGoogleSingIn.', () => { 
        // console.log(store.getState());
        
        render(
            <Provider  store={ store }>
                <MemoryRouter>
                    <LoginPage />
                </MemoryRouter>
            </Provider>
        );

        const button = screen.getByLabelText('google-btn');
        fireEvent.click( button );
        // console.log(store.getState());

        expect( mockStartGoogleSignIn ).toHaveBeenCalled();
    });

    test('Submit debe de llamar el startLoginEmailPassword.', () => { 
        const email = 'marisol@gmail.com';
        const password = '123456';

        render(
            <Provider store={ store }>
                <MemoryRouter>
                    <LoginPage />
                </MemoryRouter>
            </Provider>
        );

        const emailField = screen.getByRole('textbox', { name: 'Correo' });
        fireEvent.change( emailField, { target: { name: 'email', value: email } } );
        
        const passwordField = screen.getByTestId('password');
        fireEvent.change( passwordField, { target: { password: 'password', value: password } } );
        
        const loginForm = screen.getByLabelText('submit-form');
        fireEvent.submit( loginForm );

        expect( mockStartLoginEmailPassword ).toHaveBeenCalled();
        expect( mockStartLoginEmailPassword ).toHaveBeenCalledWith({
            email: email,
            password: password
        });
    })
});