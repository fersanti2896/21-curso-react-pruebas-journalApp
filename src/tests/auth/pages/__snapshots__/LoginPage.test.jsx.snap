// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <LoginPage /> Debe hacer match con el snapshot 1`] = `
<div>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column css-ro1yt6-MuiGrid-root"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3 box-shadow css-ixm7zf-MuiGrid-root"
    >
      <h5
        class="MuiTypography-root MuiTypography-h5 css-1jcm1rz-MuiTypography-root"
      >
        Iniciar Sesión
      </h5>
      <form
        aria-label="submit-form"
        class="animate__animated animate__fadeIn animate__faster"
      >
        <div
          class="MuiGrid-root MuiGrid-container css-11lq3yg-MuiGrid-root"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-yp0tso-MuiGrid-root"
          >
            <div
              class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="false"
                for=":r2:"
                id=":r2:-label"
              >
                Correo
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-describedby=":r2:-helper-text"
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                  id=":r2:"
                  name="email"
                  placeholder="correo@google.com"
                  type="email"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-yjsfm1"
                  >
                    <span>
                      Correo
                    </span>
                  </legend>
                </fieldset>
              </div>
              <p
                class="MuiFormHelperText-root MuiFormHelperText-sizeMedium MuiFormHelperText-contained css-1wc848c-MuiFormHelperText-root"
                id=":r2:-helper-text"
              >
                El correo debe tener un @.
              </p>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-yp0tso-MuiGrid-root"
          >
            <div
              class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
            >
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-describedby=":r3:-helper-text"
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                  data-testid="password"
                  id=":r3:"
                  name="password"
                  placeholder="Contraseña"
                  type="password"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-ihdtdm"
                  >
                    <span
                      class="notranslate"
                    >
                      ​
                    </span>
                  </legend>
                </fieldset>
              </div>
              <p
                class="MuiFormHelperText-root MuiFormHelperText-sizeMedium MuiFormHelperText-contained css-1wc848c-MuiFormHelperText-root"
                id=":r3:-helper-text"
              >
                La contraseña debe tener más de 6 caracteres.
              </p>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-l8v5xp-MuiGrid-root"
          >
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1ptpg7m-MuiGrid-root"
            >
              <div
                class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiAlert-root MuiAlert-standardError MuiAlert-standard css-1hdyuqk-MuiPaper-root-MuiAlert-root"
                role="alert"
              >
                <div
                  class="MuiAlert-icon css-1ytlwq5-MuiAlert-icon"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
                    data-testid="ErrorOutlineIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M11 15h2v2h-2zm0-8h2v6h-2zm.99-5C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                    />
                  </svg>
                </div>
                <div
                  class="MuiAlert-message css-1pxa9xg-MuiAlert-message"
                />
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 css-1b3l6lk-MuiGrid-root"
            >
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth css-1fu7jd5-MuiButtonBase-root-MuiButton-root"
                tabindex="0"
                type="submit"
              >
                <p
                  class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
                >
                  Login
                </p>
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 css-1b3l6lk-MuiGrid-root"
            >
              <button
                aria-label="google-btn"
                class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth css-1fu7jd5-MuiButtonBase-root-MuiButton-root"
                tabindex="0"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="GoogleIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12.545,10.239v3.821h5.445c-0.712,2.315-2.647,3.972-5.445,3.972c-3.332,0-6.033-2.701-6.033-6.032s2.701-6.032,6.033-6.032c1.498,0,2.866,0.549,3.921,1.453l2.814-2.814C17.503,2.988,15.139,2,12.545,2C7.021,2,2.543,6.477,2.543,12s4.478,10,10.002,10c8.396,0,10.249-7.85,9.426-11.748L12.545,10.239z"
                  />
                </svg>
                <p
                  class="MuiTypography-root MuiTypography-body1 css-yyvcc2-MuiTypography-root"
                >
                  Google
                </p>
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-container css-1svu6jh-MuiGrid-root"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 css-1fbouff-MuiTypography-root"
            >
              ¿No tienes cuenta?
            </p>
            <a
              class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-1cghntr-MuiTypography-root-MuiLink-root"
              href="/auth/register"
            >
              Crear una cuenta
            </a>
          </div>
        </div>
      </form>
    </div>
  </div>
</div>
`;
